### PR TITLE
Fix #19 Save development mode and optimal settings when CF credentials are valid, otherwise display API credentials error

### LINKED ISSUE #19: Cloudflare error while saving Optimal Settings
**Reproduce the issue** ✅ 
Reproduced on my local test site

**Identify the root cause** ✅ 
The root cause is related to the order of the operations. 
Optimal Settings are saved with AJAX call, while Cloudflare credentials are saved when submitting WP Rocket settings page.

**Scope a solution** ✅ 
The development mode and optimal settings options should not be updated with AJAX. They need valid, secure credentials before being switched on/off, to make sure we perform the CF API calls as expected.

**Estimate the effort** ✅ 
Depending on the solution selected the effort should be [XS] if Solution 1 is selected or [S] if solution 2 is selected.

**Acceptance Criteria**

- Enable Cloudflare add-on, enter valid credentials
- Enable `Optimal Settings`
- Click `Save Changes`
- No error message should appear. And Optimal settings should be applied correctly (check with Cloudflare account manually). The Optimal settings notice should appear. 

**What If**

Couldn't think of edge cases. Suggestions are welcome

### PR DESCRIPTION
**Pull request by crystinutzaa**
Part of: #20 & https://github.com/wp-media/wp-rocket/issues/2374

@hellofromtonya I cannot run any Tests because I get an error

### CODE REVIEW

**Comment by crystinutzaa** [comment]
Yes it would be nice. However I am not sure how we can get rid of the conditionals in there.

I need to know when Cloudflare settings were updated: CF credentials or development mode and optimal settings are changed. So, there are a lot of conditionals.

So, I'm bailing out if credentials were changed or settings were updated & credentials are invalid. 

What I can do is to create 2 new methods for cloudflare_devmode & cloudflare_auto_settings. 

Any other ideas? Probably my head is too deep into this one.

**Comment by Tabrisrp** [comment]
Yes start with that, create new methods for `cloudflare_devmode` and `cloudflare_auto_settings`.

I think `cloudflare_auto_settings` can also be split into smaller methods too after, one for each setting.

File: Subscriber.php [hunk 1]
```diff
@@ -370,15 +372,29 @@ public function save_cloudflare_options( $old_value, $value ) {
...
+		}
+
+		if ( ( isset( $old_value['cloudflare_devmode'], $value['cloudflare_devmode'] ) && (int) $old_value['cloudflare_devmode'] !== (int) $value['cloudflare_devmode'] ) ||
+			( isset( $old_value['cloudflare_auto_settings'], $value['cloudflare_auto_settings'] ) && (int) $old_value['cloudflare_auto_settings'] !== (int) $value['cloudflare_auto_settings'] ) ) {
+			$submit_cloudflare_view = true;
+		}
+
+		// Revalidate Cloudflare credentials if transient is false.
+		if ( false === $is_api_keys_valid_cloudflare ) {
+			$is_api_keys_valid_cloudflare = $this->cloudflare->is_api_keys_valid( $value['cloudflare_email'], $value['cloudflare_api_key'], $value['cloudflare_zone_id'] );
```
Comment (Screenfeed) [inline Subscriber.php]: Same suggestion as in #2374.
```suggestion
			if ( isset( $value['cloudflare_email'], $value['cloudflare_api_key'], $value['cloudflare_zone_id'] ) ) {
				$is_api_keys_valid_cloudflare = $this->cloudflare->is_api_keys_valid( $value['cloudflare_email'], $value['cloudflare_api_key'], $value['cloudflare_zone_id'] );
			} else {
				$is_api_keys_valid_cloudflare = false;
			}
```

File: Subscriber.php [hunk 2]
```diff
@@ -347,6 +347,168 @@ public function maybe_print_update_settings_notice() {
...
 
+	/**
+	 * Save Cloudflare dev mode admin option.
+	 *
+	 * @since 3.5.2
+	 * @author Soponar Cristina
+	 *
+	 * @param array $devmode New value for Cloudflare dev mode.
+	 */
+	private function save_cloudflare_devmode( $devmode ) {
```
Comment (Screenfeed) [inline Subscriber.php]: If the argument is an array, it should be type-hinted.
```suggestion
	private function save_cloudflare_devmode( array $devmode ) {
```
Same goes for the other methods.

File: Subscriber.php [hunk 3]
```diff
@@ -347,6 +347,168 @@ public function maybe_print_update_settings_notice() {
...
+	 * Save Cloudflare auto settings admin option.
+	 *
+	 * @since 3.5.2
+	 * @author Soponar Cristina
+	 *
+	 * @param array $auto_settings New value for Cloudflare auto_settings.
+	 * @param array $old_settings  Cloudflare cloudflare_old_settings.
+	 */
+	private function save_cloudflare_auto_settings( $auto_settings, $old_settings ) {
+		$cf_old_settings          = explode( ',', $old_settings );
```
Comment (Screenfeed) [inline Subscriber.php]: The DocBlock says `$old_settings` is an array.

File: Subscriber.php [hunk 4]
```diff
@@ -347,6 +347,168 @@ public function maybe_print_update_settings_notice() {
...
+		$cloudflare_update_result = [];
+
+		// Set Cache Level to Aggressive.
+		$cf_cache_level             = isset( $cf_old_settings[0] ) && 0 === $auto_settings ? 'basic' : 'aggressive';
```
Comment (Screenfeed) [inline Subscriber.php]: The DocBlock says `$auto_settings` is an array.

File: Subscriber.php (hunk 2 above)
Comment (crystinutzaa) [reply Subscriber.php]: Forgot to update the docblock. It's done now

Comment (Screenfeed) [reply Subscriber.php]: #CopyPasteDrama 😏

Comment (crystinutzaa) [reply Subscriber.php]: It was from the first refactor and then it ended up in the #CopyPasteDrama 🥇
